README.md
![BVox](http://bvox.net/images/logo-bvox-big.png)
# knife-xenserver

Provision virtual machines with Citrix XenServer and Opscode Chef.

## Upgrading knife-xenserver

When upgrading knife-xenserver, it's very important to remove older knife-xenserver versions

    gem update knife-xenserver
    gem clean knife-xenserver

## Usage

    knife xenserver --help

## Examples

List all the VMs

    knife xenserver vm list --xenserver-host fooserver \
                            --xenserver-username root \
                            --xenserver-password secret


List custom templates

    knife xenserver template list --xenserver-host fooserver \
                                  --xenserver-username root \
                                  --xenserver-password secret

Include built-in templates too

    knife xenserver template list --xenserver-host fooserver \
                                  --xenserver-username root \
                                  --xenserver-password secret \
                                  --include-builtin

Create a new template from a VHD file (PV by default, use --hvm otherwise) 

    knife xenserver template create --vm-name ubuntu-precise-amd64 \
                                    --vm-disk ubuntu-precise.vhd \
                                    --vm-memory 512 \
                                    --vm-networks 'Integration-VLAN' \
                                    --storage-repository 'Local storage' \
                                    --xenserver-password changeme \
                                    --xenserver-host 10.0.0.2 


Create a VM from template ed089e35-fb49-f555-4e20-9b7f3db8df2d and bootstrap it using the 'root' user and password 'secret'. The VM is created without VIFs, inherited VIFs from template are removed by default (use --keep-template-networks to avoid that)

    knife xenserver vm create --vm-template ed089e35-fb49-f555-4e20-9b7f3db8df2d \
                              --vm-name foobar --ssh-user root \
                              --ssh-password secret 

Create a VM from template and add two custom VIFs in networks 'Integration-VLAN' and 'Another-VLAN', with MAC address 11:22:33:44:55:66 for the first VIF

    knife xenserver vm create --vm-template ed089e35-fb49-f555-4e20-9b7f3db8df2d \
                              --vm-name foobar --ssh-user root \
                              --ssh-password secret \
                              --vm-networks 'Integration-VLAN,Another-VLAN' \
                              --mac-addresses 11:22:33:44:55:66

Create a VM from template and supply ip/host/domain configuration. Requires installation of xe-automater scripts
(https://github.com/krobertson/xenserver-automater)

    knife xenserver vm create   --vm-template my-template -x root --keep-template-networks \
                                --vm-name my-hostname \
                                --vm-ip 172.20.1.25 --vm-netmask 255.255.0.0 --vm-gateway 172.20.0.1 --vm-dns 172.20.0.3 \
                                --vm-domain my-domain.local

The domU/guest will also need xe-guest-utilities installed. You can then list xenstore attributes running 'xenstore-ls vm-data' inside domU.

List hypervisor networks

    knife xenserver network list

## Sample .chef/knife.rb config

    knife[:xenserver_password] = "secret"
    knife[:xenserver_username] = "root"
    knife[:xenserver_host]     = "xenserver-real"


# Building the rubygem

    gem build knife-xenserver.gemspec


CHANGELOG.md
# 1.3.2 - 18 jun 2013 

* Fix --sr-type option parsing for 'sr list' command. Thanks to @elvin159.

# 1.3.1 - Thy 06 Jun 2013 

* Fixing some --extra-vdis parsing issues
* Fix nasty bug while trying to roll back destroying the created VM

# 1.3.0 - Thy 06 Jun 2013 

* New **sr list** command to list storage repositories.
* Experimental **sr create** command to create a Storage Repository.
* The **vm create** command now supports --extra-vdis option to
  create additional VDIs and attach them to the VM.
* New **host list** command: currently lists UUID only.


# 1.2.3 - Fri 07 Dec 2012

* Drop alchemist gem, fixing some compat issues with some ruby 1.9
  implementations. See #6.

# 1.2.2 - Thu 06 Dec 2012 

* Fixed --storage-repository parameter in 'template create' command

# 1.2.1 - Mon 12 Nov 2012

* Maks3w patches for xenserver-automated related stuff

* Added --match flag to the 'vm delete' command.

  If --match is used, every VM matching VM_NAME will be deleted

* Deprecate --force-delete option in 'vm delete' since --yes should
  be used to confirm (force).

# 1.2 - Thu 25 Oct 2012 

* Greatly improved 'vm list' command

  - Table columns can now be disabled in output:
  
      knife xenserver vm list --no-uuid \
                              --no-power \
                              --no-networks \
                              --no-tools 
  
  - Print CSV output with --csv instead of a regular ASCII table
    
    knife xenserver vm list --csv

* Added --match option to 'vm list'

  Print only VMs whose name matches the given regex:

    knife xenserver vm list --match '^my-vm.*?devel.bvox.net$'

* FIX: Set exit status to 1 when XenServer auth fails
* FIX: Print error if XenServer host is not defined

# 1.1 - 2012/10/21

* Fixed --no-host-key-verify vm create flag
* added power on/off commands
* Allow setting of network information on create for Ubuntu systems 
  (@krobertson, @adamlau)

  See https://github.com/bvox/knife-xenserver/pull/1

# 0.1.0 - 2012/04/04

* Initial release


lib/chef/knife/xenserver_vm_create.rb
#
# Author:: Sergio Rubio (<rubiojr@bvox.net>)
# Copyright:: Copyright (c) 2012 BVox S.L.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'chef/knife/xenserver_base'
require 'singleton'

class Chef
  class Knife
    class XenserverVmCreate < Knife

      include Knife::XenserverBase

      deps do
        require 'readline'
        require 'chef/json_compat'
        require 'chef/knife/bootstrap'
        Chef::Knife::Bootstrap.load_deps
      end

      banner "knife xenserver vm create (options)"

      option :vm_template,
        :long => "--vm-template NAME",
        :description => "The Virtual Machine Template to use"
      
      option :vm_name,
        :long => "--vm-name NAME",
        :description => "The Virtual Machine name"
      
      option :vm_tags,
        :long => "--vm-tags tag1[,tag2..]",
        :description => "Comma separated list of tags"
      
      option :chef_node_name,
        :short => "-N NAME",
        :long => "--node-name NAME",
        :description => "The Chef node name for your new node"
      
      option :vm_memory,
        :long => "--vm-memory AMOUNT",
        :description => "The memory limits of the Virtual Machine",
        :default => '512'
      
      option :vm_cpus,
        :long => "--vm-cpus AMOUNT",
        :description => "The VCPUs of the Virtual Machine",
        :default => '1'

      option :bootstrap_version,
        :long => "--bootstrap-version VERSION",
        :description => "The version of Chef to install",
        :proc => Proc.new { |v| Chef::Config[:knife][:bootstrap_version] = v }

      option :distro,
        :short => "-d DISTRO",
        :long => "--distro DISTRO",
        :description => "Bootstrap a distro using a template; default is 'ubuntu10.04-gems'",
        :proc => Proc.new { |d| Chef::Config[:knife][:distro] = d },
        :default => "ubuntu10.04-gems"

      option :template_file,
        :long => "--template-file TEMPLATE",
        :description => "Full path to location of template to use",
        :proc => Proc.new { |t| Chef::Config[:knife][:template_file] = t },
        :default => false

      option :run_list,
        :short => "-r RUN_LIST",
        :long => "--run-list RUN_LIST",
        :description => "Comma separated list of roles/recipes to apply",
        :proc => lambda { |o| o.split(/[\s,]+/) },
        :default => []

      option :ssh_user,
        :short => "-x USERNAME",
        :long => "--ssh-user USERNAME",
        :description => "The ssh username; default is 'root'",
        :default => "root"
      
      option :ssh_password,
        :short => "-P PASSWORD",
        :long => "--ssh-password PASSWORD",
        :description => "The ssh password"

      option :identity_file,
        :short => "-i IDENTITY_FILE",
        :long => "--identity-file IDENTITY_FILE",
        :description => "The SSH identity file used for authentication"
      
      option :host_key_verify,
        :long => "--[no-]host-key-verify",
        :description => "Disable host key verification",
        :boolean => true,
        :default => true
      
      option :skip_bootstrap,
        :long => "--skip-bootstrap",
        :description => "Skip bootstrap process (Deploy only mode)",
        :boolean => true,
        :default => false,
        :proc => Proc.new { true }
      
      option :keep_template_networks,
        :long => "--keep-template-networks",
        :description => "Do no remove template inherited networks (VIFs)",
        :boolean => true,
        :default => false,
        :proc => Proc.new { true }
      
      option :batch,
        :long => "--batch script.yml",
        :description => "Use a batch file to deploy multiple VMs",
        :default => nil
      
      option :vm_networks,
        :short => "-N network[,network..]",
        :long => "--vm-networks",
        :description => "Network where nic is attached to"

      option :mac_addresses,
        :short => "-M mac[,mac..]",
        :long => "--mac-addresses",
        :description => "Mac address list",
        :default => nil

      option :vm_ip,
        :long => '--vm-ip IP',
        :description => 'IP address to set in xenstore'

      option :vm_gateway,
        :long => '--vm-gateway GATEWAY',
        :description => 'Gateway address to set in xenstore'

      option :vm_netmask,
        :long => '--vm-netmask NETMASK',
        :description => 'Netmask to set in xenstore'

      option :vm_dns,
        :long => '--vm-dns NAMESERVER',
        :description => 'DNS servers to set in xenstore'

      option :vm_domain,
        :long => '--vm-domain DOMAIN',
        :description => 'DOMAIN of host to set in xenstore'

      option :extra_vdis,
        :long => '--extra-vdis "SR name":size1[,"SR NAME":size2,..]',
        :description => 'Create and attach additional VDIs (size in MB)'

      def tcp_test_ssh(hostname)
        tcp_socket = TCPSocket.new(hostname, 22)
        readable = IO.select([tcp_socket], nil, nil, 5)
        if readable
          Chef::Log.debug("sshd accepting connections on #{hostname}, banner is #{tcp_socket.gets}")
          yield
          true
        else
          false
        end
      rescue Errno::ETIMEDOUT, Errno::EPERM
        false
      rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH, Errno::ENETUNREACH
        sleep 2
        false
      ensure
        tcp_socket && tcp_socket.close
      end

      def run
        $stdout.sync = true

        unless config[:vm_template]
          raise "You have not provided a valid template name. (--vm-template)"
        end
        
        vm_name = config[:vm_name]
        if not vm_name
          raise "Invalid Virtual Machine name (--vm-name)"
        end

        unless config[:vm_gateway]
          if config[:vm_ip]
            last_octet = Chef::Config[:knife][:xenserver_default_vm_gateway_last_octet] || 1
            config[:vm_gateway] = config[:vm_ip].gsub(/\.\d+$/, ".#{last_octet}")
          end
        end

        config[:vm_netmask] ||= Chef::Config[:knife][:xenserver_default_vm_netmask]
        config[:vm_dns] ||= Chef::Config[:knife][:xenserver_default_vm_dns]
        config[:vm_domain] ||= Chef::Config[:knife][:xenserver_default_vm_domain]

        template = connection.servers.templates.find do |s| 
          (s.name == config[:vm_template]) or (s.uuid == config[:vm_template])
        end

        if template.nil?
          raise "Template #{config[:vm_template]} not found."
        end

        puts "Creating VM #{config[:vm_name].yellow}..."
        puts "Using template #{template.name.yellow} [uuid: #{template.uuid}]..."
        
        vm = connection.servers.new :name => config[:vm_name],
                                    :template_name => config[:vm_template]
        vm.save :auto_start => false
        # Useful for the global exception handler
        @created_vm = vm

        if not config[:keep_template_networks]
          vm.vifs.each do |vif|
            vif.destroy
          end 
          vm.reload
        end
        if config[:vm_networks]
          create_nics(config[:vm_networks], config[:mac_addresses], vm)
        end
        mem = (config[:vm_memory].to_i * 1024 * 1024).to_s
        vm.set_attribute 'memory_limits', mem, mem, mem, mem
        vm.set_attribute 'VCPUs_max', config[:vm_cpus]
        vm.set_attribute 'VCPUs_at_startup', config[:vm_cpus]

        # network configuration through xenstore
        attrs = {}
        (attrs['vm-data/ip'] = config[:vm_ip]) if config[:vm_ip]
        (attrs['vm-data/gw'] = config[:vm_gateway]) if config[:vm_gateway]
        (attrs['vm-data/nm'] = config[:vm_netmask]) if config[:vm_netmask]
        (attrs['vm-data/ns'] = config[:vm_dns]) if config[:vm_dns]
        (attrs['vm-data/dm'] = config[:vm_domain]) if config[:vm_domain]
        if !attrs.empty?
          puts "Adding attributes to xenstore..."
          vm.set_attribute 'xenstore_data', attrs
        end

        if config[:vm_tags]
          vm.set_attribute 'tags', config[:vm_tags].split(',')
        end

        vm.provision
        # Create additional VDIs (virtual disks)
        create_extra_vdis(vm)
        vm.start
        vm.reload

        puts "#{ui.color("VM Name", :cyan)}: #{vm.name}"
        puts "#{ui.color("VM Memory", :cyan)}: #{bytes_to_megabytes(vm.memory_static_max)} MB"

        if !config[:skip_bootstrap]
          # wait for it to be ready to do stuff
          print "\n#{ui.color("Waiting server... ", :magenta)}"
          timeout = 180
          found = connection.servers.all.find { |v| v.name == vm.name }
          servers = connection.servers
          if config[:vm_ip]
            vm.refresh
            print "\nTrying to #{'SSH'.yellow} to #{config[:vm_ip].yellow}... "
            print(".") until tcp_test_ssh(config[:vm_ip]) do
              sleep @initial_sleep_delay ||= 10; puts(" done")
              @ssh_ip = config[:vm_ip]
            end
          else
            loop do
              begin
                vm.refresh
                if not vm.guest_metrics.nil? and not vm.guest_metrics.networks.empty?
                  networks = []
                  vm.guest_metrics.networks.each do |k,v|
                    networks << v
                  end
                  networks = networks.join(",")
                  puts
                  puts "\n#{ui.color("Server IPs:", :cyan)} #{networks}"
                  break
                end
              rescue Fog::Errors::Error
                print "\r#{ui.color('Waiting a valid IP', :magenta)}..." + "." * (100 - timeout)
              end
              sleep 1
              timeout -= 1
              if timeout == 0
                puts
                raise "Timeout trying to reach the VM. Couldn't find the IP address."
              end
            end
            print "\n#{ui.color("Waiting for sshd... ", :magenta)}"
            vm.guest_metrics.networks.each do |k,v|
              print "\nTrying to #{'SSH'.yellow} to #{v.yellow}... "
              print(".") until tcp_test_ssh(v) do
                sleep @initial_sleep_delay ||= 10; puts(" done")
                @ssh_ip = v
              end
              break if @ssh_ip
            end
          end

          bootstrap_for_node(vm).run 
          puts "\n"
          puts "#{ui.color("Name", :cyan)}: #{vm.name}"
          puts "#{ui.color("IP Address", :cyan)}: #{@ssh_ip}"
          puts "#{ui.color("Environment", :cyan)}: #{config[:environment] || '_default'}"
          puts "#{ui.color("Run List", :cyan)}: #{config[:run_list].join(', ')}"
          puts "#{ui.color("Done!", :green)}"
        else
          ui.warn "Skipping bootstrapping as requested."
        end

      end

      def bootstrap_for_node(vm)
        bootstrap = Chef::Knife::Bootstrap.new
        bootstrap.name_args = [@ssh_ip]
        bootstrap.config[:run_list] = config[:run_list]
        bootstrap.config[:ssh_user] = config[:ssh_user] 
        bootstrap.config[:identity_file] = config[:identity_file]
        bootstrap.config[:chef_node_name] = config[:chef_node_name] || vm.name
        bootstrap.config[:bootstrap_version] = locate_config_value(:bootstrap_version)
        bootstrap.config[:distro] = locate_config_value(:distro)
        # bootstrap will run as root...sudo (by default) also messes up Ohai on CentOS boxes
        bootstrap.config[:use_sudo] = true unless config[:ssh_user] == 'root'
        bootstrap.config[:template_file] = locate_config_value(:template_file)
        bootstrap.config[:environment] = config[:environment]
        bootstrap.config[:host_key_verify] = config[:host_key_verify]
        bootstrap.config[:ssh_password] = config[:ssh_password]
        bootstrap
      end

      def create_extra_vdis(vm)
        # Return if no extra VDIs were specified
        return unless config[:extra_vdis]
        count = 0

        vdis = config[:extra_vdis].strip.chomp.split(',')
        vdis.each do |vdi|
          count += 1
          # if options is "Storage Repository":size
          if vdi =~ /.*:.*/
            sr, size = vdi.split(':')
          else #only size was specified
            sr = nil
            size = vdi
          end
          unless size =~ /^\d+$/
            raise "Invalid VDI size. Not numeric."
          end
          # size in bytes
          bsize = size.to_i * 1024 * 1024

          # If the storage repository has been omitted,
          # use the default SR from the first pool, othewise
          # find the SR required
          if sr.nil?
            sr = connection.pools.first.default_sr 
            ui.warn "No storage repository defined for extra VDI #{count}."
            ui.warn "Using default SR from Pool: #{sr.name}"
          else
            found = connection.storage_repositories.find { |s| s.name == sr }
            unless found
              raise "Storage Repository #{sr} not available"
            end
            sr = found
          end
          
          # Name of the VDI
          name = "#{config[:vm_name]}-extra-vdi-#{count}"

          puts "Creating extra VDI (#{size} MB, #{name}, #{sr.name || 'Default SR'})"
          vdi = connection.vdis.create :name => name,
                                       :storage_repository => sr,
                                       :description => name,
                                       :virtual_size => bsize.to_s

          begin
            # Attach the VBD
            connection.vbds.create :server => vm, 
                                   :vdi => vdi,
                                   :userdevice => count.to_s,
                                   :bootable => false
          rescue => e
            ui.error "Could not attach the VBD to the server"
            # Try to destroy the VDI
            vdi.destroy rescue nil
            raise e
          end
        end
      end
      
      def create_nics(networks, macs, vm)
        net_arr = networks.split(/,/).map { |x| { :network => x } }
        nics = []
        if macs
          mac_arr = macs.split(/,/)
          nics = net_arr.each_index { |x| net_arr[x][:mac_address] = mac_arr[x] if mac_arr[x] and !mac_arr[x].empty? }
        else
          nics = net_arr
        end
        networks = connection.networks
        highest_device = -1
        vm.vifs.each { |vif| highest_device = vif.device.to_i if vif.device.to_i > highest_device }
        nic_count = 0
        nics.each do |n|
          net = networks.find { |net| net.name == n[:network] }
          if net.nil?
            raise "Network #{n[:network]} not found"
          end
          nic_count += 1
          c = {
           'MAC_autogenerated' => n[:mac_address].nil? ? 'True':'False',
           'VM' => vm.reference,
           'network' => net.reference,
           'MAC' => n[:mac_address] || '',
           'device' => (highest_device + nic_count).to_s,
           'MTU' => '0',
           'other_config' => {},
           'qos_algorithm_type' => 'ratelimit',
           'qos_algorithm_params' => {}
          }
          connection.create_vif_custom c
        end
      end

    end
  end
end
